test_runner/src/main/kotlin/ftl/run/common/FetchArtifacts.kt

```diff
@@ -1,76 +1,50 @@
 package ftl.run.common
 
-import com.google.cloud.storage.Storage
 import flank.common.OutputLogLevel
 import flank.common.log
 import flank.common.logLn
 import flank.common.startWithNewLine
+import ftl.api.Artifacts.DownloadPath
+import ftl.api.Artifacts.Identity
+import ftl.api.fetchArtifacts
 import ftl.args.IArgs
-import ftl.client.google.GcStorage
 import ftl.config.FtlConstants
 import ftl.json.MatrixMap
 import ftl.util.Artifacts
 import ftl.util.MatrixState
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.Job
 import kotlinx.coroutines.coroutineScope
-import kotlinx.coroutines.joinAll
-import kotlinx.coroutines.launch
-import java.nio.file.Path
-import java.nio.file.Paths
 
-// TODO needs refactor
-internal suspend fun fetchArtifacts(matrixMap: MatrixMap, args: IArgs) = coroutineScope {
+internal suspend fun fetchAllTestRunArtifacts(matrixMap: MatrixMap, args: IArgs) = coroutineScope {
     logLn("FetchArtifacts".startWithNewLine(), OutputLogLevel.DETAILED)
-    val fields = Storage.BlobListOption.fields(Storage.BlobField.NAME)
-
-    var dirty = false
-    val filtered = matrixMap.map.values.filter {
-        val finished = it.state == MatrixState.FINISHED
-        val notDownloaded = !it.downloaded
-        finished && notDownloaded
-    }.toMutableList()
-
     log(FtlConstants.indent)
-    filtered.flatMapIndexed { index, matrix ->
-        val prefix = Storage.BlobListOption.prefix(matrix.gcsPathWithoutRootBucket)
-        val result = GcStorage.storage.list(matrix.gcsRootBucket, prefix, fields)
-        val artifactsList = Artifacts.regexList(args)
-        val jobs = mutableListOf<Job>()
 
-        result.iterateAll().forEach { blob ->
-            val blobPath = blob.blobId.name
-            val matched = artifactsList.any { blobPath.matches(it) }
-            if (matched) {
-                val downloadFile = getDownloadPath(args, blobPath)
-                if (!downloadFile.toFile().exists()) {
-                    val parentFile = downloadFile.parent.toFile()
-                    parentFile.mkdirs()
-                    jobs += launch(Dispatchers.IO) { blob.downloadTo(downloadFile) }
-                }
-            }
+    val downloadPath = DownloadPath(
+        localResultDir = args.localResultDir,
+        useLocalResultDir = args.useLocalResultDir(),
+        keepFilePath = args.keepFilePath
+    )
+
+    matrixMap.map.values
+        .filter {
+            val finished = it.state == MatrixState.FINISHED
+            val notDownloaded = !it.downloaded
+            finished && notDownloaded
+        }
+        .map {
+            Identity(
+                gcsPathWithoutRootBucket = it.gcsPathWithoutRootBucket,
+                gcsRootBucket = it.gcsRootBucket,
+                regex = Artifacts.regexList(args),
```

Comment:
```suggestion
    val artifactRegexList = Artifacts.regexList(args)

    matrixMap.map.values
        .filter {
            val finished = it.state == MatrixState.FINISHED
            val notDownloaded = !it.downloaded
            finished && notDownloaded
        }
        .map {
            Identity(
                gcsPathWithoutRootBucket = it.gcsPathWithoutRootBucket,
                gcsRootBucket = it.gcsRootBucket,
                regex = artifactRegexList,
```

The value returned by `Artifacts.regexList(args)` is immutable so it can be reused over all the all identities

Reply:
oh nice catch, thanks!